api/src/main/java/net/kyori/adventure/text/Component.java

```diff
@@ -143,20 +145,91 @@
    * @param components the components
    * @return a text component
    * @since 4.0.0
+   * @deprecated for removal since 4.8.0, use {@link #join(JoinConfiguration, Iterable)} with a {@link JoinConfiguration} instead.
    */
+  @Deprecated
   @Contract(value = "_, _ -> new", pure = true)
   static @NonNull TextComponent join(final @NonNull ComponentLike separator, final Iterable<? extends ComponentLike> components) {
+    final Component component = join(JoinConfiguration.separator(separator), components);
+
+    if(component instanceof TextComponent) return (TextComponent) component;
+    return Component.text().append(component).build();
+  }
+
+  /**
+   * Joins {@code components} using the configuration in {@code config}.
+   *
+   * @param config the join configuration
+   * @param components the components
+   * @return a text component
+   * @see JoinConfiguration
+   * @since 4.8.0
+   */
+  @Contract(pure = true)
+  static @NonNull Component join(final @NonNull JoinConfiguration config, final @NonNull ComponentLike@NonNull... components) {
+    return join(config, Arrays.asList(components));
+  }
+
+  /**
+   * Joins {@code components} using the configuration in {@code config}.
+   *
+   * @param config the join configuration
+   * @param components the components
+   * @return a text component
+   * @see JoinConfiguration
+   * @since 4.8.0
+   */
+  @Contract(pure = true)
+  static @NonNull Component join(final @NonNull JoinConfiguration config, final @NonNull Iterable<? extends ComponentLike> components) {
     final Iterator<? extends ComponentLike> it = components.iterator();
+    final Component prefix = config.prefix();
```

Comment:
This logic is becoming more complex. What about moving the implementation into `JoinConfigurationImpl`?